Feature: verify the authenticity of a webhook event

stefanvermaas opened the pull request:
Fixes #11819 

---

When Spree sends a webhook event to a webhook subscriber, there is no way for the webhook subscriber to verify the authenticity of the webhook event. Due to this, a webhook subscriber needs to allow only certain IP-addresses to be sure the webhook event is actually coming from a verified source.

This PR adds support for an event signature (based on the subscriber's `secret_key` and the payload), so the webhook subscriber can verify the authenticity of the source of the event.

The event signature is added as an HTTP request header (`X-Spree-Hmac-SHA256` or `HTTP_X_SPREE_HMAC_SHA256` for PHP, or a Rack-based framework such as Ruby on Rails or Sinatra) for every request send to the webhook subscriber. 

**It's up to the webhook subscriber to implement the verification of the authenticity.**

Hereby an example (taken from Shopify's documentation):

```ruby
# The following example uses Ruby and the Sinatra web framework to verify a webhook request:
require 'rubygems'
require 'base64'
require 'openssl'
require 'sinatra'
require 'active_support/security_utils'

# The webhook secret key is viewable from the Spree Backend Dashboard. 
# In a production environment, it's recommended to set the client secret as an environment variable to prevent exposing it in code.
CLIENT_SECRET = 'my_client_secret'

helpers do
  # Compare the computed HMAC digest based on the client secret and the request contents to the reported HMAC in the headers
  def verify_webhook(data, hmac_header)
    calculated_hmac = Base64.strict_encode64(OpenSSL::HMAC.digest('sha256', CLIENT_SECRET, data))
    ActiveSupport::SecurityUtils.secure_compare(calculated_hmac, hmac_header)
  end
end

# Respond to HTTP POST requests sent to this web service
post '/' do
  request.body.rewind
  data = request.body.read
  verified = verify_webhook(data, env["HTTP_X_SPREE_HMAC_SHA256"])

  halt 401 unless verified

  # Process webhook payload
  # ...
end
```

Feedback I'm looking for:

- Is the addition of event's signature in the `Spree::Webhooks::Subscribers::MakeRequest` implemented in a way that is conform the code style of the Spree project?
- The chosen name of the HTTP header is `X-Spree-Hmac-SHA256`. Do you feel this is a suiting name?
- The visibility of the `WebhookSubscriber#secret_key` in the admin dashboard isn't implemented yet as I wanted to do that in a separate PR. Do you agree?

I haven't updated the documentation (`spree-dev-docs`) yet, but I also noticed there is no documentation yet for the webhooks. Is any help needed on that front?

Looking forward to your feedback. Merry Christmas 🎄

stefanvermaas committed:
Adds the secret_key column to the spree_webhooks_subscribers table

stefanvermaas committed:
Adds the presence validation for the new secret_key

stefanvermaas committed:
Adds automatic generation of a secret key for webhook subscribers

stefanvermaas committed:
Adds Spree::Webhooks::EventSignature to compute a unique signature

stefanvermaas committed:
Add the event's signature to the request

stefanvermaas committed:
Forward the event's signature in the HandleRequest

stefanvermaas committed:
Fixes a typo in the event_spec.rb

stefanvermaas committed:
Adds Spree::Webhooks::Event#signature_for

The #signature_for method adds a shorthand method for computing the event's signature, so the receiver of the payload can confirm the authenticity of the webhook event

stefanvermaas committed:
Fixes failing specs after adding the secret key to the subscribers

stefanvermaas commented:
PS: Do I need to update something to be able to run the CI on Circle?

rafalcymerys commented:
Hi @stefanvermaas this looks great at first sight, thank you for sending this over. I will do a more in-depth review and will come back with more feedback then.

To answer your questions really quick:
1. _Is the addition of event's signature in the Spree::Webhooks::Subscribers::MakeRequest implemented in a way that is conform the code style of the Spree project?_ 
At first sight, it looks really clean

2. _The chosen name of the HTTP header is X-Spree-Hmac-SHA256. Do you feel this is a suiting name?_ 
I think it's fine. @damianlegawiec @nciemniak @kshalot any thoughts on that?

3. _The visibility of the WebhookSubscriber#secret_key in the admin dashboard isn't implemented yet as I wanted to do that in a separate PR. Do you agree?_ 
Yes, also given that the admin panel is a separate repository, this makes total sense.

When it comes to the docs, we could definitely do better when it comes to documenting webhooks. 
If you feel like it, please create a PR in this repository https://github.com/spree/spree-dev-docs . Your change will be definitely worth mentioning there, as this is a useful feature. We're also working on restructuring the docs a bit, to make it easier to find information, so if you have any thoughts on what would be the best section to put it in, let us know.

nciemniak commented:
> The chosen name of the HTTP header is X-Spree-Hmac-SHA256. Do you feel this is a suiting name?
> I think it's fine. @damianlegawiec @nciemniak @kshalot any thoughts on that?

Agree, name sounds good to me 👍

kshalot commented:
Just a thought - we could also opt for a more generic header, i.e. one that does not contain a specific hash function in the name? AFAIR Shopify uses `X-Shopify-Hmac-SHA256` and Slack `X-Slack-Signature` so both are used in the wild. I guess the advantage is that the name is decoupled from the hash function which might have some pros and cons. That being said, I'm okay with the proposed name.

On a different note - we should probably also include the timestamp in a header, e.g. `X-Spree-Timestamp`/`X-Spree-Request-Timestamp`, and also include it in the signature. This will prevent events from being valid forever.

rafalcymerys commented:
@kshalot +1 for including a timestamp - not necessarily as a part of this PR, but it's a common practice that goes along with request signatures.

stefanvermaas commented:
All right, thanks all for the feedback so far. Any thoughts on getting the CI up and running? What do I need to do differently?

Looking forward to any more in-depth feedback.

stefanvermaas commented on api/app/models/spree/webhooks/subscriber.rb:
```diff
@@ -17,6 +18,7 @@ class Subscriber < Spree::Webhooks::Base
       scope :active, -> { where(active: true) }
       scope :inactive, -> { where(active: false) }
 
+      before_validation :generate_secret_key, unless: :secret_key?
```
**NOTE:** The only thing I didn't take care of is the webhook subscribers which already exist. Those won't have a secret key as we speak. 

I think it makes sense to add some "helper methods" which can do things such as; regenerating a secret key and/or generating a secret key for all existing subscribers, etc.

I'm not sure about doing it in this PR, but I can do it in a follow-up PR. How has the Spree team handled these kinda migrations in the past?

rafalcymerys replied:
Yes, the usual way to go would be to create a migration that will backfill/generate the data for existing records. 
They get copied into the project when someone upgrades Spree, so this serves this exact purpose

rafalcymerys commented:
> All right, thanks all for the feedback so far. Any thoughts on getting the CI up and running? What do I need to do differently?

Looking forward to any more in-depth feedback.

@stefanvermaas are you able to sign in to circleci with your github account, and then rerun the tests on your branch? Since it's your fork, I think it needs you to authenticate with CircleCI.

rafalcymerys commented on api/app/models/spree/webhooks/subscriber.rb:
```diff
@@ -51,6 +53,10 @@ def check_uri_path
       end
 
+      def generate_secret_key
+        assign_attributes(secret_key: SecureRandom.urlsafe_base64(32))
```
Is there a side-effect that we rely on here, or could it be simplified to `self.secret_key = SecureRandom.urlsafe_base64(32)`?

stefanvermaas replied:
No, there is no side effect here for the `secret_key` attribute in this case. Normally, I use the `assign_attributes` as it runs the "entire" lifecycle of attribute assignment. 

I'm fine with whatever you see fit here. If it's more common to use the `self.secret_key = ...`, I'm more than happy to change it.

rafalcymerys replied:
I just did a grep across the codebase, and we don't use assign_attributes in other places - it's it's always a direct assignment using `self.xxx = yyy`. If you could update that, it will be consistent with the rest of the codebase then. Thanks!

rafalcymerys commented on api/db/migrate/20221221122100_add_secret_key_to_spree_webhooks_subscribers.rb:
```diff
@@ -0,0 +1,5 @@
+class AddSecretKeyToSpreeWebhooksSubscribers < ActiveRecord::Migration[7.0]
+  def change
+    add_column :spree_webhooks_subscribers, :secret_key, :string, null: true
```
I think `OpenSSL::HMAC.digest('sha256', @event.subscriber.secret_key, payload)` will fail when the key parameter is `nil`. I'd suggest adding two more migrations as a part of this PR:
1. A migration that will generate and backfill empty secret_keys for spree_webhooks_subscribers
2. A migration that will disallow nulls in the secret_key column

This way we'll make sure we won't break any webhook subscribers in existing installations of Spree

stefanvermaas replied:
Thank you! I'll add these migrations. As I said before, I wasn't fully aware of how the Spree core team likes to handle these kinds of migrations. I'll make the adjustments.

rafalcymerys replied:
Thanks!

rafalcymerys reviewed (commented):
@stefanvermaas I took a detailed look and this looks great. I left two comments - one that I think is quite important for compatibility reasons, and a minor one related to assign_attributes vs self.attr = . Let me know your thoughts on these.

Did you have any luck with spinning up the tests on circleci? I can try to run them on Spree's account if needed.

stefanvermaas commented:
> Did you have any luck with spinning up the tests on circleci? I can try to run them on Spree's account if needed.

I have not. I can run them for my own fork, but I don't seem to be able to run them in the PR. Would you mind trying to kick them off through the Spree Circle CI account? When I click on the "Details" in the Github Actions, it also brings me to the Spree Circle CI account it seems.

rafalcymerys commented:
@stefanvermaas when it comes to the CircleCI issues, I'll need to investigate them further. CircleCI recently had a security incident, so maybe they rotated OAuth tokens. 

Anyway, I just ran your branch on our CI manually, waiting for the build to pass: https://app.circleci.com/pipelines/github/spree/spree/6242/workflows/b4adfc0b-e592-4557-b27c-f8bb03fd2f75

stefanvermaas committed:
Use self.secret_key= over assign_attributes(secret_key: ...)

stefanvermaas committed:
Simplify the generation of a secret key by depending on Rails' has_secure_token

stefanvermaas committed:
Add a migration to backfill webhook subscribers with a missing secret key

stefanvermaas committed:
Mark the secret key column as non-null by default

stefanvermaas commented:
@rafalcymerys I've added the migrations. 

I've also added a notable change: we're now depending on Rails' `has_secure_token` instead of a custom implementation as that also gave us `regenerate_secret_key` by default. It's a Rails 6+ feature, so I don't expect any problems here.

I'm also thinking of removing the specs [here](https://github.com/spree/spree/pull/11821/files#diff-9abb8e0a97efc271dc736a49a4945ae873127b1b08775e4d75e1ee32b669c21fR22-R43) as we're now testing default Rails behavior and I don't think we should test for standard behavior. What do you think?

Last but not least, I see the specs are failing, but they don't seem to be related to any changes I've made. Are you aware of this?

rafalcymerys commented:
Hi @stefanvermaas - thank you! 

_I've also added a notable change: we're now depending on Rails' has_secure_token instead of a custom implementation as that also gave us regenerate_secret_key by default. It's a Rails 6+ feature, so I don't expect any problems here._

Yeah, that's fine - we only support Rails 6.1+ now anyway, so no issues here.

_I'm also thinking of removing the specs [here](https://github.com/spree/spree/pull/11821/files#diff-9abb8e0a97efc271dc736a49a4945ae873127b1b08775e4d75e1ee32b669c21fR22-R43) as we're now testing default Rails behavior and I don't think we should test for standard behavior. What do you think?_

Let me review them. That would make sense as a separate PR, right?


_Last but not least, I see the specs are failing, but they don't seem to be related to any changes I've made. Are you aware of this?_

I haven't reviewed them yet, just ran the build on your branch :) I see that a lot of tests in this build are failing on: https://app.circleci.com/pipelines/github/stefanvermaas/spree/5/workflows/74f87c1e-58c7-4b48-a512-43b6e5503dff/jobs/48
```       
NoMethodError:
undefined method `secret_key?' for #<Spree::Webhooks::Subscriber:0x000055b8b18aebd8>
```


I can take a look at that, but I wanted to confirm that this is the issue you were referring to.

stefanvermaas commented:
> Let me review them. That would make sense as a separate PR, right?

Yeah, sure. It can be a separate PR. 

> I haven't reviewed them yet, just ran the build on your branch :) I see that a lot of tests in this build are failing on: https://app.circleci.com/pipelines/github/stefanvermaas/spree/5/workflows/74f87c1e-58c7-4b48-a512-43b6e5503dff/jobs/48


Hmm, that's interesting. I thought I saw a different error, but this one seems to be related. I'll double-check and come back to this.

stefanvermaas commented:
Ah – I was correct. There were other errors I was seeing and you were also right, there are specs failing in relation to this PR. 

These are the failing specs, but I can't replicate them locally. When I run them, they're just turning out fine. Have you seen this before? 

```shell
API v2 Caching spec serializers caching auto expiration auto expire cache after record being updated
     Failure/Error: expect(json_response['data']).to have_attribute('name').with_value('Some name')
       expected `name` attribute to have value `Some name` but was ``

API V2 Storefront Products Spec products#index with specified options with no locale set returns products with specified options
     Failure/Error: expect(json_response['data'].first).to have_id(product_with_option.id.to_s)
     
     NoMethodError:
       undefined method `[]' for nil:NilClass
```

I'm not 100% sure why we're seeing the error you've pasted (I see them too with Rails 6.1), so I assume the `?` method addition wasn't added yet in Rails 6.1. I'll adjust the code a bit for consistency to double-check.

stefanvermaas committed:
Address the missing #secret_key? in Rails 6.x

stefanvermaas committed:
Correctly spell existence

rafalcymerys commented:
Hi @stefanvermaas ! There are a few tests that fail randomly sometimes. I think it's mostly dependent on the order in which they ran - we've recently fixed a few spots where e.g. a single tests changed i18n.locale, but didn't reset it afterwards, which sometimes broke tests that ran after that one. I will rerun them on our circleci workers and will see if that helps.

rafalcymerys commented:
Other than that, your PR looks awesome and I'll merge it as soon as we figure out what's going on with the tests

rafalcymerys committed:
Merge remote-tracking branch 'origin/main' into feature/webhook-authenticity-verification

rafalcymerys commented:
@stefanvermaas I re-ran the test suite on your PR (also merged the latest main to it, since we made some fixes to how libvips is installed). 

https://app.circleci.com/pipelines/github/spree/spree/6310/workflows/bf026370-e655-40f5-a1fb-200bd73cb62d
They seem to pass fine, except the suite for Rails 6.1. It seems that the `secret_key` method that's called by the `secret_key?` that you've implemented, is not available for some reason. But that's a column on the webhook subscribers table, right?

stefanvermaas commented:
> They seem to pass fine, except the suite for Rails 6.1. It seems that the secret_key method that's called by the secret_key? that you've implemented, is not available for some reason. But that's a column on the webhook subscribers table, right?

Yeah, I'm not sure why that happens. It's a regular column. See this migration that adds it; https://github.com/spree/spree/pull/11821/files#diff-811c3dd2107cb4a817702ed8f9bcc297e7e53f87ced337a66b71e76045874e54R3

Maybe it's a caching issue on CircleCI? Could we run the tests again with the caches being cleared? 

I'm also relatively sure we actually don't need the `#secret_key?` method as it should have been defined by Rails natively. I'll remove it, I'll push it and if you can re-run the test suite without any cache, we can say more about why it fails.

stefanvermaas committed:
Removes the Spree::Webhooks::Subscriber#secret_key? definition

rafalcymerys commented:
Hi @stefanvermaas - I think I found the issue: in the migration files that you've added, could you change the migration version from `7.0` to `6.1`? The current setup makes the migrations fail when running with Rails 6.1, and for some reason they fail silently in CI. That's why Rails is unable to see this column in tests and they fail :) 

It solved the issue for me locally, I also ran a build with this fix to double check: https://app.circleci.com/pipelines/github/spree/spree/6334/workflows/c8455c7e-cc1d-49d8-9596-3a42a097c729

stefanvermaas committed:
Downgrade migrations to Rails 6.1

stefanvermaas commented:
> Hi @stefanvermaas - I think I found the issue: in the migration files that you've added, could you change the migration version from `7.0` to `6.1`? The current setup makes the migrations fail when running with Rails 6.1, and for some reason they fail silently in CI. That's why Rails is unable to see this column in tests and they fail :)
> 
> It solved the issue for me locally, I also ran a build with this fix to double check: https://app.circleci.com/pipelines/github/spree/spree/6334/workflows/c8455c7e-cc1d-49d8-9596-3a42a097c729

Oh gosh. You're right. That's a great catch. I never thought about that 🤦 . Just pushed the changes. Let's see how they run!

rafalcymerys commented:
Hi @stefanvermaas I don't know why the tests failed in your pipeline, but I ran your branch here: https://app.circleci.com/pipelines/github/spree/spree/6346/workflows/182ce3c1-0ece-42f2-afff-cf75994ab7d0 and they passed fine :) 

Thank you so much for this PR, merging it now!

stefanvermaas commented:
Thank you @rafalcymerys for your patience and guidance!
